internal/pgadmin/users.go

```diff
@@ -0,0 +1,128 @@
+)
+
+type Executor func(
+	ctx context.Context, stdin io.Reader, stdout, stderr io.Writer, command ...string,
+) error
+
+// WriteUsersInPGAdmin uses exec and "python3" to create users in pgAdmin and
+// update their passwords when they already exist. A blank password for a user
+// blocks that user from logging in to pgAdmin. The pgAdmin configuration
+// database must exist before calling this.
+func WriteUsersInPGAdmin(
+	ctx context.Context, exec Executor,
+	users []v1beta1.PostgresUserSpec, passwords map[string]string,
+) error {
+	// The location of pgAdmin files can vary by container image. Look for
+	// typical names in the module search path: the PyPI package is named
+	// "pgadmin4" while custom builds might use "pgadmin4-web". The pgAdmin
+	// packages expect to find themselves on the search path, so prepend that
+	// directory there (like pgAdmin does in its WSGI entrypoint).
+	// - https://pypi.org/project/pgadmin4/
+	// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/pgAdmin4.wsgi;hb=REL-4_20#l13
+	const search = `
+import importlib.util, os, sys
+
+spec = importlib.util.find_spec('pgadmin4') or importlib.util.find_spec('pgadmin4-web')
+root = os.path.dirname(importlib.util.find_spec('.pgadmin', spec.name).submodule_search_locations[0])
+if sys.path[0] != root: sys.path.insert(0, root)
+`
+
+	// The user with id=1 is automatically created by pgAdmin when it creates
+	// its configuration database. Clear that username so it cannot conflict
+	// with the users we create, and deactivate the user so it cannot log in.
+	// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/migrations/versions/fdc58d9bd449_.py;hb=REL-4_20#l129
+	//
+	// The "verify_and_update_password" method hashes the plaintext password
+	// according to pgAdmin security settings. It is part of the User model
+	// since pgAdmin v4.19 and Flask-Security-Too v3.20.
+	// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=requirements.txt;hb=REL-4_20#l40
+	// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/pgadmin/model/__init__.py;hb=REL-4_20#l65
+	// - https://flask-security-too.readthedocs.io/en/stable/api.html#flask_security.UserMixin.verify_and_update_password
+	//
+	// TODO(cbandy): pgAdmin v4.21 adds "auth_source" and "username" as required attributes.
+	// - https://git.postgresql.org/gitweb/?p=pgadmin4.git;f=web/pgadmin/model/__init__.py;hb=REL-4_21#l65
+	const script = `
+import json, sys
+from pgadmin import create_app
+from pgadmin.model import db, Role, User
+
+with create_app().app_context():
+  admin = db.session.query(User).filter_by(id=1).first()
+  admin.active = False
+  admin.email = ''
+  admin.password = ''
+
+  db.session.add(admin)
+  db.session.commit()
+
+  for line in sys.stdin:
+    if not line.strip(): continue
+
+    data = json.loads(line)
+    user = (
+      db.session.query(User).filter_by(email=data['username']).first() or
```

Comment:
`data['username']` -- if for whatever reason this is blank, this will throw a `KeyError`.

If you want to fail softly, I'd suggest:

```suggestion
      db.session.query(User).filter_by(email=data.get('username')).first() or
```

Reply:
et al. for the other ones.

Reply:
By "blank" you mean missing from the dictionary, right?

```pycon
>>> data = dict(empty='', full='asdf')
>>> data['empty']
''
>>> data['full']
'asdf'
>>> data['other']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'other'
```

🤔 I think an exception here is okay. It describes the problem and the line that caused it, and the `python` command exits non-zero. What else might we do if the input JSON is incorrect?

Reply:
Yes I meant missing.

We could gracefully handle the failure or if we keep it as a KeyError, provide a more informative message.

Reply:
I do feel as though the current behavior/exception is exactly what we're looking for, i.e. we get the non-zero exit with an error that describes the problem well enough.  I do also like keeping these embedded scripts as terse as possible.

Therefore, not seeing the need for any additional error handling here at this time.